### ISSUE TITLE
New Package Request: google-authenticator-libpam

### ISSUE THREAD

**Issue by tymonx** [coreos/fedora-coreos-tracker#1037]
Project: [google-authenticator-libpam](https://github.com/google/google-authenticator-libpam)

Fedora Magazine article: [Set up two-factor authentication for SSH on Fedora](https://fedoramagazine.org/two-factor-authentication-ssh-fedora/)

Please try to answer the following questions about the package you are requesting:

1. What, if any, are the additional dependencies on the package? (i.e. does it pull in Python, Perl, etc)

No additional dependencies. It is a pure C source code that is normally compiling using any existing C compiler with PAM headers. It uses the same dependencies like any other existing PAM modules. All required dynamic libraries are already present in the Fedora CoreOS:

- `linux-vdso.so.1`
- `libpam.so.0`
- `libdl.so.2`
- `libc.so.6`
- `libaudit.so.1`
- `libeconf.so.0`
- `libm.so.6`
- `/lib64/ld-linux-x86-64.so.2`
- `libcap-ng.so.0`

2. What is the size of the package and its dependencies?

For the `x86_64` architecture without any fancy build options around **129kB**.

3. What problem are you trying to solve with this package? Or what functionality does the package provide?

Two-factor authorization for remote SSH access. This will greatly improve security for deployment OS.

4. Can the software provided by the package be run from a container?  Explain why or why not.

No. It is strictly configured using PAM configuration files configured by system administrator (or using Butane/Ignition).

5. Can the tool(s) provided by the package be helpful in debugging container runtime issues?

No.

6. Can the tool(s) provided by the package be helpful in debugging networking issues?

No.

7. Is it possible to layer the package onto the base OS as a day 2 operation?  Explain why or why not.

Yes. It is quite easy to compile and install under `/usr/lib64/security` location.

8. In the case of packages providing services and binaries, can the packaging be adjusted to just deliver binaries?

Yes. It requires only compiled `pam_*.so` library and install under `/usr/lib64/security` location.

9. Can the tool(s) provided by the package be used to do things we’d rather users not be able to do in FCOS? (e.g. can it be abused as a Turing complete interpreter?)

I don't believe so.

10. Does the software provided by the package have a history of CVEs?

Yes like [CVE-2012-6140](https://www.cvedetails.com/cve/CVE-2012-6140/).

**Comment by travier** [comment]
> 7. Is it possible to layer the package onto the base OS as a day 2 operation?  Explain why or why not.
> 
> Yes. It is quite easy to compile and install under `/usr/lib64/security` location.

To clarify here, this question is about layering the package via rpm-ostree install, either on first boot or post installation. There is a [package in Fedora](https://src.fedoraproject.org/rpms/google-authenticator) so this is a Yes.

But as far as I understand this requires manual / coordinated setup for it to be usable thus you will not be able to configure it directly via Ignition. This makes it less interesting for inclusion in the base image as this requires manual attention anyway.

**Comment by tymonx** [comment]
I think the best solution would be to include some kind of `pam_otp.so` library (`pam_google_authenticator.so`) with the Fedora CoreOS base image.

This will allow to have two-factor authentication out-of-box. This is important in deployment environments with limited access to internet or any kind of external network but remote SSH access is still required for management and maintenance.

This will allow to configure it with Butane/Ignition by simple changing the `/etc/ssh/sshd_config` configuration file and the `/etc/pam.d/sshd` content.

And it will be possible to extend current Butane/Ignition specification with the `otp` field:

```yaml
passwd:
    users:
        - name: core
          otp: true # Enable two-factor authentication OTP (One-Time Password)
          ssh_authorized_keys:
              - <key>
```

**Comment by travier** [comment]
To setup OTP, you need a shared secret between the client and server. Thus this will at least need another field to store this secret.

**Comment by lucab** [comment]
Perhaps a bit of a subjective opinion, but overall I don't think we want to push *COS user towards more customized/heavyweight SSH usage.
To the contrary, we'd like to keep host-wide privileged SSH access to the minimum required for disaster discovery (and be happy if you can get rid of SSH service at all).
If a user ends up in a situation where they need both interactive shell access and heavily-customized authN/authZ mechanisms, maybe they should consider some dedicated solutions like https://goteleport.com/.

**Comment by tymonx** [comment]
@travier Yes, I'm aware of that. For now this can be handled with Butane `storage.files`.

@lucab Maybe to illustrate my request I will describe from my experience how most of my deployments looks:

- all deployments are 100% remote on machines and infrastructure manage by external customer
- network policy is very strict, no internet access, isolated, you cannot provide any additional services. Mostly of time allowing only remote IPMI or SSH access (through VPN) by IT even with harsh negotiation which ports/protocols to open
- because of that strict environment SSH (or SSH over IPMI SOL) is using for loading/updating images `docker --host ssh://<ip> load` to remote machines or updating Docker Swarm stack. Unfortunately it is not only used for disaster discovery :/ SSH is the easiest way to convince IT department why at least it is required

And I was thinking how to secure this more when someone expose/loose his SSH private key. Key rotation and revocation can be hard because it requires constant access to network and some customers automatically revoke accesses over time. TOTP works offline.

**Comment by dustymabe** [comment]
We discussed this in the [community meeting today](https://lists.fedoraproject.org/archives/list/coreos@lists.fedoraproject.org/message/HFAD6TSELVURTAVGTJHT46HFXI4IA5FU/).

```
12:10:05   dustymabe | #agreed We will not add google-authenticator-libpam for now.
                     | Overlaying the package on first-boot is the currently
                     | recommended approach. It should be investigated if SSSD would be
                     | able to perform 2FA.
```

That being said there were some other points of discussion during the meeting that are worth bringing up:

- It was suggested during the meeting that SSSD, which we already ship, may be able to handle this use case already.
- Can anyone show us how you would provision multiple nodes using Ignition and take advantage of this PAM module? 
    - i.e. getting the shared secret onto the nodes and configuring PAM to use the module. 
    - also would you be using the same shared key across all N nodes, or a different key per node?

**Comment by tymonx** [comment]
if I'm not mistaken it requires some external centralized provider like FreeIPA, LDAP or Kerberos? I have looked into sssd code and I saw only OTP abstraction to these providers but any kind of OTP implementation. Other sssd services like files, simple, sudo don't provide any kind of OTP support. But I'm not expert in sssd. Still I don't have any possibility to setup full realm domain.

How about pre-installing `/usr/lib64/security/pam_oath.so` from the existing `pam_oath` package? And how to resolve chicken-egg problem for SSH+OTP without OTP after OS installation? Without external internet access to install it with the ostree-rpm? Go with CoreOS Assembler and custom image?

**Comment by dustymabe** [comment]
On the first question:

> - It was suggested during the meeting that SSSD, which we already ship, may be able to handle this use case already.

I talked to a few experts on this:

> @sgallagher:
> I know that SSSD supports two-factor authentication when used with
> FreeIPA (which supports both TOTP and HOTP authenticators). I don't
> know if there's a general implementation, though. CCing Sumit Bose to
> comment.

> @sumit-bose
> I'm sorry there is no general OTP implementation in SSSD.

@tymonx - you are right.

**Comment by dustymabe** [comment]
On the second question:

> - Can anyone show us how you would provision multiple nodes using Ignition and take advantage of this PAM module?
>     - i.e. getting the shared secret onto the nodes and configuring PAM to use the module.
>     - also would you be using the same shared key across all N nodes, or a different key per node?

@tymonx if I gave you a dev build with this package included could you show us exactly how you would use it? i.e. what Ignition config you would use to deploy everything with no manual interaction?

**Comment by tymonx** [comment]
I can create a public git sample project with Butane YAML to show how to setup SSH + OTP for Fedora CoreOS. I already have some little experience with custom derived Fedora CoreOS and I can prepare configuration that will include preinstalled OTP PAM module (`google-authenticator-libpam` or `pam_aoth`). With instruction how to test it with `qemu` or `libvirt`.

I can also prepare an example for [Fedora CoreOS Documentation](https://docs.fedoraproject.org/en-US/fedora-coreos) to show how to setup and use SSH with OTP. This should be a nice security add-on for everyone who like to harden their deployments.
